### PR TITLE
linenoise: init at git-2016-09-30

### PR DESCRIPTION
**Pull request by fstamour**
###### Motivation for this change

I needed that library in order to build another custom package. Might be useful for others.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
  - [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) AFAIK.

---

There is place for amelioration: static and dynamic builds and corresponding pkg-config file.

### CODE REVIEW

File: pkgs/development/libraries/linenoise/default.nix
Comment (joachifm) [inline pkgs/development/libraries/linenoise/default.nix]: Please use the convention `<name>-unstable-<version>` or simply `<name>-<version>`. Note that version strings must begin with a digit, everything before the first `-<digit>` is parsed as part of the name.

File: pkgs/development/libraries/linenoise/default.nix
Comment (joachifm) [inline pkgs/development/libraries/linenoise/default.nix]: Please disable problematic phases instead; while this may work in this case, it is poor practice.

File: pkgs/development/libraries/linenoise/default.nix
Comment (joachifm) [inline pkgs/development/libraries/linenoise/default.nix]: Please specify a license and preferably add yourself as the package maintainer.

File: pkgs/development/libraries/linenoise/default.nix
Comment (joachifm) [inline pkgs/development/libraries/linenoise/default.nix]: Please use `fetchFromGitHub` instead; archives are generally preferred over checkouts.

**Review by joachifm — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/linenoise/default.nix
Comment (mpsyco) [inline pkgs/development/libraries/linenoise/default.nix]: How do I do that?

Comment (joachifm) [inline pkgs/development/libraries/linenoise/default.nix]: Which phase is problematic? The default phases are supposed to be somewhat smart wrt. not doing things unless appropriate files are found in the build toplevel.

For some phases there are flags, like `dontBuild = true` or you can override the phase to a noop, as in `buildPhase = ":"`. See https://nixos.org/nixpkgs/manual/#sec-stdenv-phases for details.

Comment (mpsyco) [inline pkgs/development/libraries/linenoise/default.nix]: Oh, I didn't know about the noop.
So you prefer that I explicitly says that each phases are noop rather than saying that only those phases are used?

File: pkgs/development/libraries/linenoise/default.nix
Comment (mpsyco) [inline pkgs/development/libraries/linenoise/default.nix]: Done

File: pkgs/development/libraries/linenoise/default.nix
Comment (mpsyco) [inline pkgs/development/libraries/linenoise/default.nix]: Is it ok now?

File: pkgs/development/libraries/linenoise/default.nix
Comment (joachifm) [inline pkgs/development/libraries/linenoise/default.nix]: Ideally, you'd only disable phases if they actually fail.

The problem with explicitly listing phases, in general, is that you can miss out on useful phases. This can cause issues with massive closure bloat or run-time crashes. Again, this is more of a general point, I'm sure it won't make that much of a difference in this specific case.

**Review by fstamour — COMMENTED** [review]
I also changed fetchgit -> fetchFromGithub.

File: pkgs/development/libraries/linenoise/default.nix
Comment (mpsyco) [inline pkgs/development/libraries/linenoise/default.nix]: Yeah, based on the ouput I can see that if someone added the "static lib build" version of linenoise, it would a lot of binary manipulations.

Anyway, it's changed!

**Comment by mpsyco** [comment]
@joachifm Based on your feedback, everything should be all right now.

File: pkgs/development/libraries/linenoise/default.nix
Comment (joachifm) [inline pkgs/development/libraries/linenoise/default.nix]: I see you're using both `stdenv.lib` and `lib`, perhaps stick to the former for consistency (they should be the same thing).

**Review by joachifm — APPROVED** [review]

**Comment by joachifm** [comment]
Having quickly looked over the changed version, looks pretty good to me. Thank you.

File: pkgs/development/libraries/linenoise/create-pkg-config-file.sh
Comment (joachifm) [inline pkgs/development/libraries/linenoise/create-pkg-config-file.sh]: Also, this is out of sync with the expression. Maybe generate this inline or use `substituteAll` to ensure consistency.

**Comment by joachifm** [comment]
Actually there were two more minor issues. Also, once you're ready to merge, everything needs to be squashed into a single commit first.

File: pkgs/development/libraries/linenoise/default.nix
Comment (FRidh) [inline pkgs/development/libraries/linenoise/default.nix]: if I am correct you can use `rev = version;`

Comment (FRidh) [inline pkgs/development/libraries/linenoise/default.nix]: or at least, use the tag they used for the version

**Comment by grahamc** [comment]
Nice work, @mpsyco.

File: pkgs/development/libraries/linenoise/default.nix
Comment (mpsyco) [inline pkgs/development/libraries/linenoise/default.nix]: The thing is that I kind of invented the version number. The only tag on the repo is "1.0", but it's 10 commits away of `head`, hence the `1.0.10` version.I hope it's ok with you?

**Comment by mpsyco** [comment]
Ok, I (manually, not using `substituteAll` yet) made the pkg-config's version and the nix package's version matches (good catch btw) and I removed the redundant `lib` and I squashed everything.

This is the first time I squash commits and I know I won't remember how, so just so I can retrace how I did it: [here is the StackOverflow I used](http://stackoverflow.com/a/5201642/1564493).

**Comment by mpsyco** [comment]
@grahamc Thanks, and thanks for your help on IRC :smiley: .

**Comment by joachifm** [comment]
Merged. Thank you

File: pkgs/development/libraries/linenoise/create-pkg-config-file.sh
Comment (edolstra) [inline pkgs/development/libraries/linenoise/create-pkg-config-file.sh]: Having a C source in `Cflags` strikes me as odd. Is that even allowed? Wouldn't this lead to linenoise.c being compiled into every compilation unit that uses the Cflags from linenoise.pc?

Comment (fstamour) [inline pkgs/development/libraries/linenoise/create-pkg-config-file.sh]: Yes, the library will be compiled with the project that uses it. Yes, it is permitted. No, it's not nice.

I wanted to make a second pull request that provides at least a static library and at best a dynamically linked one. But I didn't even used that library since I added it to nixpkgs.

**Pull request closed without merging**
